storage: move to pkg/kv/kvserver

tbg opened the pull request:
```
pkg/kv/kvserver <-- today: pkg/storage
pkg/storage     <-- today: pkg/storage/engine
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45640)
<!-- Reviewable:end -->

tbg commented:
Heh, I botched the rename (I used kvserver in one commit, storage in the other) but I just wanted to see if CI can become green anyway.

knz committed:
cli: prevent `demo` and `start` from working if TZ db not avail

Previously, unavailability of named time zones (issue #36864) was only
causing an error message to be printed on the terminal and the log
files. However, some users were not looking and would instead run
directly into silently incorrect results in SQL.

This change ensures a node doesn't even start when the TZ database is
not ready.

Example output:

```
ERROR: unable to load named timezones
HINT: See: https://github.com/cockroachdb/cockroach/issues/36864
--
Check that the time zone database is installed on your system, or
set the ZONEINFO environment variable to a Go time zone .zip archive.
Failed running "demo"
```

Release note (cli change): CockroachDB now refuses to start if named
time zones are not properly configured. It is possible to override
this behavior for testing purposes, with the understanding that doing
so will cause incorrect SQL results and other inconsistencies, using
the environment variable `COCKROACH_INCONSISTENT_TIME_ZONES`.

tbg committed:
storage: git mv to pkg/kv/storage

The final location will be pkg/kv/kvserver but this intermediate step
helps isolate the change of the package name.

This commit does not compile. No manual edits were made.

Release note: None

tbg committed:
storage: Goland-move to kv/storage

No manual edits. Does not compile (proto import paths etc).
I created this (and follow-up commits of the same type) by
first reverting the parent commit (`git mv ...`), doing the
refactor via the IDE, and squashing the revert into this commit.

Release note: None

tbg committed:
storage: git mv pkg/kv/{storage,kvserver}

No manual edits. Won't compile.

Release note: None

tbg committed:
storage: Goland-rename to pkg/kv/kvserver

Release note: None

tbg committed:
kv/kvserver/engine: git mv to ./engine

Final name will be `storage`. Does not compile.

Release note: None

tbg committed:
engine: Goland-move to top-level pkg/engine

Final name will be ./pkg/storage. No manual edits.

Release note: None

tbg committed:
engine: git mv to storage

No manual edits.

Release note: None

tbg committed:
engine: Goland-rename to storage

No manual edits.

Release note: None

tbg committed:
enginepb: fix import/usage paths

storage/engine -> storage
storage.engine -> storage
storage::engine -> storage

Release note: None

tbg committed:
kvserver: fix proto paths

```
search:  storage(.*)(storagepb|ctupdate|concurrency|copysets)
replace: kv$1kvserver$1$2
```

No manual edits.

Release note: None

tbg committed:
libroach: fix up CMakelists.txt

Release note: None

tbg committed:
Makefile: use new storage package

Release note: None

tbg committed:
storage: fix CPPFLAGS

Release note: None

tbg committed:
kvserver: fix proto package name

`make protobuf` now works.

Release note: None

tbg committed:
Makefile: engine/enginepb -> enginepb

Release note: None

tbg committed:
*: more engine/enginepb replacements

Release note: None

tbg committed:
kv: more proto package name changes

Release note: None

tbg committed:
*: regenerate protos

`make build` now works.

Release note: None

knz committed:
util/log: make SecondaryLogger a Closer and do the right thing

Previously, secondary loggers would leak in the global registry.
A previous patch of mine tried to improve the situation by
asynchronously de-registering loggers, but that was not good enough.
This patch fixes that by adding a `Close()` method to do the job.

Release note: None

knz committed:
util/log: make SecondaryLogger a Closer and do the right thing

Previously, secondary loggers would leak in the global registry.
A previous patch of mine tried to improve the situation by
asynchronously de-registering loggers, but that was not good enough.
This patch fixes that by adding a `Close()` method to do the job.

Release note: None

tbg committed:
github-pull-request-make: don't stress failed tests in this PR

This is a clear hack, but worth it to get this PR merged.

Release note: None

tbg committed:
kv,storage: fix go:generate invocations

Fixes `make generate`.

Release note: None

tbg committed:
github-post: fix TestListFailures

Release note: None

tbg committed:
lint: fix excluded paths

Release note: None

tbg committed:
fixup! github-pull-request-make: flip the bool

Messed this up before.

Release note: None

tbg committed:
storage,kvserver: fix doc comments

Release note: None

tbg committed:
ui: fix proto references

Release note: None

nvanbenschoten commented on pkg/testutils/lint/lint_test.go:
```diff
@@ -1018,7 +1018,7 @@ func TestLint(t *testing.T) {
 			filter,
 			stream.GrepNot(`.*\.lock`),
-			stream.GrepNot(`^storage\/engine\/rocksdb_error_dict\.go$`),
+			stream.GrepNot(`^storage\/rocksdb_error_dict\.go$`),
```
Not for now, but what's up with this? Don't we have the same grep twice now?

nvb reviewed (approved):
But I know you have more coming down the pipeline.

tbg committed:
kvserver: fix RPC service names

Changing proto package changes service name, breaking mixed version
clusters. There's something to clean up here.

Release note: None

tbg commented:
bors r=nvanbenschoten

I ran a number of TC invocations. The last known problem was that the mixed version tests were failing, which is now fixed (verified locally). Going to try this.

tbg committed:
build: skip testrace when too many packages changed

If lots of packages changed, don't run testrace on the assumption that
this is a large-scale refactor.

NB: it would be better to use PKG=./pkg/... in this case.

Release note: None

tbg commented:
Both CI runs failed on TestSecondary, which is flaky on master and fixed in http://github.com/cockroachdb/cockroach/pull/45695. That PR is now merging, so retrying.

I also added a commit that skips testrace if more than 10 packages changed. I will revisit that when this PR merges, too, but it is necessary to avoid blocking bors for >2h.

bors r=nvanbenschoten

tbg committed:
Revert "util/log: make SecondaryLogger a Closer and do the right thing"

This reverts commit f13e05d956d8375b3f5ddb6854539195cc3ca700.

Release note (<category, see below>): <what> <show> <why>

tbg commented:
I caught a merge conflict with f13e05d956d8375b3f5ddb6854539195cc3ca700 (which is bors'ing right now and which I need to merge to get the flake out of the way).

I'm changing this PR to revert this commit first, then cherry-pick it again (including resolution of the merge conflict).

tbg commented:
bors r=nvanbenschoten

tbg commented:
bors r+

Messed something up in the "skip race" commit. Aborting this early, so much stuff will be queued in 2h

tbg commented:
The bors CI ran through (https://teamcity.cockroachdb.com/viewLog.html?buildId=1785201&buildTypeId=Cockroach_UnitTests&tab=buildResultsDiv&branch_Cockroach=staging) but then threw up because the license/cla webhook hadn't fired.
I'm calling this fair game for a manual merge.
